strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.tinkerpop.gremlin.structure.Element;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.apache.tinkerpop.gremlin.structure.VertexProperty.Cardinality;
+import org.carlspring.strongbox.data.domain.DomainObject;
+import org.carlspring.strongbox.data.domain.EntityHierarchyNode;
+import org.carlspring.strongbox.gremlin.adapters.UnfoldEntityTraversal;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * @author sbespalov
+ *
+ * @param <S>
+ * @param <E>
+ */
+@GremlinDsl(traversalSource = "org.carlspring.strongbox.gremlin.dsl.EntityTraversalSourceDsl")
+public interface EntityTraversalDsl<S, E> extends GraphTraversal.Admin<S, E>
+{
+
+    Logger logger = LoggerFactory.getLogger(EntityTraversalDsl.class);
+
+    Object NULL = new Object()
+    {
+
+        @Override
+        public String toString()
+        {
+            return "__null";
+        }
+
+    };
+
+    @SuppressWarnings("unchecked")
+    default GraphTraversal<S, Vertex> findById(Object uuid,
+                                               String... labels)
+    {
+        GraphTraversal<S, Vertex> result = (GraphTraversal<S, Vertex>) has(labels[0], "uuid", uuid);
+        for (String label : Arrays.copyOfRange(labels, 1, labels.length))
+        {
+            result = result.fold().choose(Collection::isEmpty, __.V().has(label, "uuid", uuid), __.unfold());
+        }
+
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    default Traversal<S, Object> enrichPropertyValue(String propertyName)
+    {
+        return coalesce(__.properties(propertyName).value(), __.<Object>constant(NULL));
+    }
+
+    @SuppressWarnings("unchecked")
+    default Traversal<S, Object> enrichPropertyValues(String propertyName)
+    {
+        return coalesce(__.propertyMap(propertyName).map(t -> t.get().get(propertyName)), __.<Object>constant(NULL));
+    }
+
+    default <S2> Traversal<S, Object> mapToObject(Traversal<S2, Object> enrichObjectTraversal)
+    {
+        return fold().choose(t -> t.isEmpty(),
+                             __.<Object>constant(NULL),
+                             __.<Edge>unfold().map(enrichObjectTraversal));
+    }
+
+    default GraphTraversal<S, Vertex> V(DomainObject entity)
+    {
+        if (entity instanceof EntityHierarchyNode)
+        {
+            EntityHierarchyNode entityHierarchyNode = (EntityHierarchyNode) entity;
+            while (entityHierarchyNode.getHierarchyParent() != null)
+            {
+                entityHierarchyNode = entityHierarchyNode.getHierarchyParent();
+            }
+            entity = (DomainObject) entityHierarchyNode;
+        }
+
+        Long nativeId = entity.getNativeId();
+        if (nativeId != null)
+        {
+            return ((EntityTraversal<S, Vertex>) V(nativeId)).debug("Attached");
+        }
+
+        return V();
+    }
+
+    default <S2> Traversal<S, Vertex> saveV(Object uuid,
```

Comment:
It would be make clear to name this `addOrUpdateV`, assuming I've understood it's purpose correctly?

```suggestion
    default <S2> Traversal<S, Vertex> addOrUpdateV(Object uuid,
```

Reply:
I had such naming potion for this method, but finally decided to use `saveV` because the meaning of `save` is add or update :) If you don't mind I would like leave the name like it is now because it less then `addOrUpdateV`, we could clarify the actual behavior in javadocs.